SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialVirtualWorkerTranslationManager.cpp

```diff
@@ -76,18 +76,34 @@ void SpatialVirtualWorkerTranslationManager::ConstructVirtualWorkerMappingFromQu
 			const Worker_ComponentData& Data = Entity.components[j];
 			// System entities which represent workers have a component on them which specifies the SpatialOS worker ID,
 			// which is the string we use to refer to them as a physical worker ID.
-			if (Data.component_id == SpatialConstants::WORKER_COMPONENT_ID)
+			if (Data.component_id == SpatialConstants::SERVER_WORKER_COMPONENT_ID)
 			{
 				const Schema_Object* ComponentObject = Schema_GetComponentDataFields(Data.schema_type);
 
-				const FString& WorkerType = SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_TYPE_ID);
+				// The translator should only acknowledge workers that are ready to begin play. This means we can make
+				// guarantees based on where non-GSM-authoritative servers canBeginPlayTrue=true as an AddComponent
+				// or ComponentUpdate op. This affects how startup Actors are treated in a zoned environment.
+				const bool bWorkerIsReadyToBeginPlay = SpatialGDK::GetBoolFromSchema(ComponentObject, SpatialConstants::SERVER_WORKER_READY_TO_BEGIN_PLAY_ID);
+				if (!bWorkerIsReadyToBeginPlay)
+				{
+					continue;
+				}
 
 				// TODO(zoning): Currently, this only works if server workers never die. Once we want to support replacing
 				// workers, this will need to process UnassignWorker before processing AssignWorker.
-				if (WorkerType.Equals(SpatialConstants::DefaultServerWorkerType.ToString()) &&
-					!UnassignedVirtualWorkers.IsEmpty())
+				if (!UnassignedVirtualWorkers.IsEmpty())
 				{
-					AssignWorker(SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::WORKER_ID_ID));
+					// If we didn't find all our server worker entities the first time, future query responses should
+					// ignore workers that we have already assigned a virtual worker ID.
+					const PhysicalWorkerName QueriedWorkerName = SpatialGDK::GetStringFromSchema(ComponentObject, SpatialConstants::SERVER_WORKER_NAME_ID);
+					TArray<FString> ExistingWorkerNames;
```

Comment:
Can't we just check the ExistingWorkerNames directly?

Reply:
Added a map with inverse mapping. Currently, it's just used to make this check efficient. In future, it'll be useful when we want to unassign crashed workers.